### PR TITLE
Retrieve relation information through Glue API directory

### PR DESCRIPTION
**Pull request by roy-ht**
# Problem

Our Glue database is huge and often failed to retrieve schema information.

Athena returns this type of error:
```
GENERIC_INTERNAL_ERROR: java.lang.RuntimeException: java.lang.InterruptedException: sleep interrupted
```

More specifically, 2 macros often cause an error, or query is too slow:

- https://github.com/Tomme/dbt-athena/blob/0cafe2cbc0a69a171cf8c011b7967be578c0ec1e/dbt/include/athena/macros/adapters/metadata.sql#L1
- https://github.com/Tomme/dbt-athena/blob/0cafe2cbc0a69a171cf8c011b7967be578c0ec1e/dbt/include/athena/macros/adapters/metadata.sql#L92

# Solution

Add override methods:

- `get_columns_in_relation`
- `list_relations_without_caching`
- `_get_one_catalog`

And in these methods, Call glue_get_table and glue_get_tables and get its column information directory.

# Effect

- Retrieving table schemas become faster and cache this for later use.
- when generating document by `dbt docs`, showing more table metadata information got from API like partition keys and locations.

### CODE REVIEW

File: dbt/adapters/athena/impl.py [hunk 1]
```diff
@@ -28,18 +38,51 @@ def convert_text_type(cls, agate_table: agate.Table, col_idx: int) -> str:
...
+                    database=schema_relation.database,
+                    identifier=table["Name"],
+                    schema=schema_relation.schema,
+                    quote_policy=quote_policy,
+                    type=rel_type,
+                    column_information=table["StorageDescriptor"]["Columns"],
+                )
+                relations.append(relation)
+            return relations
+        except ClientError as e:
```
Comment (roy-ht) [inline dbt/adapters/athena/impl.py:77]: When API call returns some error, Fallback into SQL implementation.
It cares some permission problems.

File: dbt/adapters/athena/impl.py [hunk 2]
```diff
@@ -48,25 +91,25 @@ def s3_uuid_table_location(self):
         return f"{client.s3_staging_dir}tables/{str(uuid4())}/"
 
     @available
-    def clean_up_partitions(
-        self, database_name: str, table_name: str, where_condition: str
-    ):
+    def clean_up_partitions(self, database_name: str, table_name: str, where_condition: str):
         # Look up Glue partitions & clean up
         conn = self.connections.get_thread_connection()
         client = conn.handle
```
Comment (roy-ht) [inline dbt/adapters/athena/impl.py:97]: dbt already has thread local connection, so we only need to use boto3 session created by PyAthena connection.

Comment (roy-ht) [reply dbt/adapters/athena/impl.py:95]: So don't have to use locks.

File: dbt/adapters/athena/impl.py [hunk 3]
```diff
@@ -53,25 +90,25 @@ def s3_uuid_table_location(self):
...
-
-        with boto3_client_lock:
-            glue_client = boto3.client('glue', region_name=client.region_name)
-        s3_resource = boto3.resource('s3', region_name=client.region_name)
+        glue_client = client.session.client("glue")
```
[the remark this answers is not in the document]
Comment (roy-ht) [reply dbt/adapters/athena/impl.py:97]: @owenprough-sift 

> Is this PR obsoleted by recently-merged #86?
No, I tried master branch that #86 was merged, it still causes an `java.lang.RuntimeException`.

**Comment by roy-ht** [comment]
@owenprough-sift 

> Is this PR obsoleted by recently-merged #86?
No, I tried master branch that #86 was merged, it still causes an `java.lang.RuntimeException`.

File: dbt/adapters/athena/impl.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (roy-ht) [reply dbt/adapters/athena/impl.py:97]: I also use `aws_profile_name` in profiles.yml :)

**Comment by roy-ht** [comment]
So uses may want to select if using Glue API or not, i'll add a configuration option like `use_glue_api: bool`.

**Comment by roy-ht** [comment]
@VDFaller 

88a868a adds the `use_glue_api` option

**Comment by Enriqson** [comment]
I also had a slowness issue with `list_relations_without_caching`, it worked fine for schemas up to 100 tables but was painfully slow for 101+.
Solved it by updating the Athena Engine to version 3.
I've tested schemas with up to 296 tables and they all perform quite well.
Guess it might help you as well @roy-ht

**Pull request closed without merging**